Docker Hub Image

bytesnz opened the pull request:
Just starting to add remote-storage (RS) to some projects I am working on, so am setting up my own RS server. I didn't see that you don't have a (recent) docker image on docker hub, so created a Dockerfile and a Github Action to build and publish one when a release is created.

The image is very basic - uses the [node:lts-alpine](https://hub.docker.com/_/node) docker image and the basic armadietto config (from `armadietto -e`). It exposes the armadietto http port to be used with a proxy.

Are you interested in including this in the repo? Figure you could create an organisation on docker hub to publish it under.

raucao commented:
Anyone reading this has enough Docker experience/knowledge to be able to review this PR? Any and all help would be most appreciated! :pray:

peponi commented:
I prefer standalone docker files, i.e. ``git clone`` within the image build and not before

reduces build steps and errors on the end user side

but this is not needed for this project IMHO
because there is **already a finished package on npmjs**

I would suggest using this

## My Dockerfile

```dockerfile
FROM alpine:latest AS appBuild

LABEL maintainer="Peponi <pep0ni@pm.com>" \
      description="will run the Armadietto NodeJS web service (a RemoteStorageJS backend)"

ARG PROJECT_NAME="armadietto"
ARG PORT="8000"
ARG PKG_MANAGER="yarn"
ARG INSTALL_COMMAND="yarn global add"
ARG CONFIG_PATH_CERTS="/etc/letsencrypt/live/example.com/"
ARG CONFIG_PATH_STORAGE="/usr/share/armadietto"

ENV PROJECT_NAME=$PROJECT_NAME

RUN set -e;\
  apk add  \
    curl \
    git \
    nodejs \
    $PKG_MANAGER; \
  mkdir /opt/armadietto; \
  mkdir -m 0700 /$CONFIG_PATH_STORAGE; \
  mkdir -m 0700 /$CONFIG_PATH_CERTS; \
  $INSTALL_COMMAND $PROJECT_NAME; \
  apk del git $PKG_MANAGER; \
  adduser -u 6582 -HD $PROJECT_NAME;

COPY config.json /etc/armadietto.conf.json

VOLUME $CONFIG_PATH_STORAGE $CONFIG_PATH_CERTS
USER $PROJECT_NAME
WORKDIR ~
EXPOSE $PORT

CMD $PROJECT_NAME -c /etc/armadietto.conf.json

HEALTHCHECK CMD curl --fail http://127.0.0.1:$PORT/ || exit 1

### Install ###
#
# BUILD:
#
# default for amd64 architecture
#
# > docker build -t armadietto:latest .
# > docker build -t --build-arg PKG_MANAGER="npm" --build-arg INSTALL_COMMAND="npm i -g" armadietto:latest .
#
# RUN:
#
# > docker run --rm -p 8000:8000 rarmadietto:latest
#
# INFO: config.json needs to be byside the Dockerfile
#
# {
#   "allow_signup": true,
#   "storage_path": "/usr/share/armadietto",
#   "cache_views": true,
#   "http": {
#     "host": "0.0.0.0",
#     "port": 8000
#   },
#   "https": {
#     "enable": false,
#     "force": false,
#     "port": 4443,
#     "cert": "/etc/letsencrypt/live/example.com/cert.pem",
#     "key": "/etc/letsencrypt/live/example.com/privkey.pem"
#   },
#   "basePath": ""
# }
#
###
```

## e2e test

![image](https://user-images.githubusercontent.com/1695282/119254651-c39a2300-bbb7-11eb-972a-946594f922d9.png)

## image size

```shell
EPOSITORY                                     TAG                 IMAGE ID            CREATED             SIZE
armadietto                                     peponi              9a84b73851eb        21 minutes ago      52.4MB
armadietto                                     bytesnz             9829e4ad797b        About an hour ago   118MB

```
## security scan with **trivy**

```shell
 ✘  ~/Code/private/DockerFiles/armadietto  docker run --rm  -v /var/run/docker.sock:/var/run/docker.sock -v cache:/root/.cache/ aquasec/trivy armadietto:peponi
2021-05-23T09:21:40.833Z        INFO    Detected OS: alpine
2021-05-23T09:21:40.833Z        INFO    Detecting Alpine vulnerabilities...
2021-05-23T09:21:40.836Z        INFO    Number of PL dependency files: 1
2021-05-23T09:21:40.836Z        INFO    Detecting yarn vulnerabilities...

armadietto:peponi (alpine 3.13.5)
=================================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)


usr/local/share/.config/yarn/global/yarn.lock
=============================================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)

```

raucao commented:
@peponi That looks great! :clap: 

@bytesnz Want to review @peponi's proposed changes to the Dockerfile?

raucao commented:
> created a Dockerfile and a Github Action to build and publish one when a release is created

I just created a remotestorage organization on Docker Hub and then created a repo for armadietto. However, I cannot find where to create custom access tokens, and the permissions tab doesn't let me use the "owners" team to create a new permission entry.

Maybe someone more experienced wants access to the RS team there? The free/community plan allows 2 more accounts to be added...

bytesnz commented:
> I prefer standalone docker files, i.e. `git clone` within the image build and not before
> 
> reduces build steps and errors on the end user side
> 
> but this is not needed for this project IMHO
> because there is **already a finished package on npmjs**
> 
> I would suggest using this
> ## My Dockerfile
> 
> ```dockerfile
> FROM alpine:latest AS appBuild
> 
> LABEL maintainer="Peponi <pep0ni@pm.com>" \
>       description="will run the Armadietto NodeJS web service (a RemoteStorageJS backend)"
> 
> ARG PROJECT_NAME="armadietto"
> ARG PORT="8000"
> ARG PKG_MANAGER="yarn"
> ARG INSTALL_COMMAND="yarn global add"
> ARG CONFIG_PATH_CERTS="/etc/letsencrypt/live/example.com/"
> ARG CONFIG_PATH_STORAGE="/usr/share/armadietto"
> 
> ENV PROJECT_NAME=$PROJECT_NAME
> 
> RUN set -e;\
>   apk add  \
>     curl \
>     git \
>     nodejs \
>     $PKG_MANAGER; \
>   mkdir /opt/armadietto; \
>   mkdir -m 0700 /$CONFIG_PATH_STORAGE; \
>   mkdir -m 0700 /$CONFIG_PATH_CERTS; \
>   $INSTALL_COMMAND $PROJECT_NAME; \
>   apk del git $PKG_MANAGER; \
>   adduser -u 6582 -HD $PROJECT_NAME;
> 
> COPY config.json /etc/armadietto.conf.json
> 
> VOLUME $CONFIG_PATH_STORAGE $CONFIG_PATH_CERTS
> USER $PROJECT_NAME
> WORKDIR ~
> EXPOSE $PORT
> 
> CMD $PROJECT_NAME -c /etc/armadietto.conf.json
> 
> HEALTHCHECK CMD curl --fail http://127.0.0.1:$PORT/ || exit 1
> 
> ### Install ###
> #
> # BUILD:
> #
> # default for amd64 architecture
> #
> # > docker build -t armadietto:latest .
> # > docker build -t --build-arg PKG_MANAGER="npm" --build-arg INSTALL_COMMAND="npm i -g" armadietto:latest .
> #
> # RUN:
> #
> # > docker run --rm -p 8000:8000 rarmadietto:latest
> #
> # INFO: config.json needs to be byside the Dockerfile
> #
> # {
> #   "allow_signup": true,
> #   "storage_path": "/usr/share/armadietto",
> #   "cache_views": true,
> #   "http": {
> #     "host": "0.0.0.0",
> #     "port": 8000
> #   },
> #   "https": {
> #     "enable": false,
> #     "force": false,
> #     "port": 4443,
> #     "cert": "/etc/letsencrypt/live/example.com/cert.pem",
> #     "key": "/etc/letsencrypt/live/example.com/privkey.pem"
> #   },
> #   "basePath": ""
> # }
> #
> ###
> ```
> 
> ## e2e test
> 
> ![image](https://user-images.githubusercontent.com/1695282/119254651-c39a2300-bbb7-11eb-972a-946594f922d9.png)
> ## image size
> 
> ```shell
> EPOSITORY                                     TAG                 IMAGE ID            CREATED             SIZE
> armadietto                                     peponi              9a84b73851eb        21 minutes ago      52.4MB
> armadietto                                     bytesnz             9829e4ad797b        About an hour ago   118MB
> ```
> 
> ## security scan with **trivy**
> 
> ```shell
>  ✘  ~/Code/private/DockerFiles/armadietto  docker run --rm  -v /var/run/docker.sock:/var/run/docker.sock -v cache:/root/.cache/ aquasec/trivy armadietto:peponi
> 2021-05-23T09:21:40.833Z        INFO    Detected OS: alpine
> 2021-05-23T09:21:40.833Z        INFO    Detecting Alpine vulnerabilities...
> 2021-05-23T09:21:40.836Z        INFO    Number of PL dependency files: 1
> 2021-05-23T09:21:40.836Z        INFO    Detecting yarn vulnerabilities...
> 
> armadietto:peponi (alpine 3.13.5)
> =================================
> Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
> 
> 
> usr/local/share/.config/yarn/global/yarn.lock
> =============================================
> Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
> ```

Cool. Looks good. Will integrate some of the stuff with the MR. The reason I went for using copy instead of yarn install is for integrating with github actions I - it would need to be published to npm before the docker image is created, but that's doable.

I think it is good to have a docker image as well as it makes it super easy to deploy.

AngeloR commented:
We may want to utilize GitHub actions to do the final build/push to docker hub as well. I'd be in favour of remaining on docker hub, but if we want we can always push the image to multiple locations.

bytesnz commented:
> We may want to utilize GitHub actions to do the final build/push to docker hub as well. I'd be in favour of remaining on docker hub, but if we want we can always push the image to multiple locations.

:+1: The MR has a template Github Action for building docker images, but it will need some tweaking.

raucao commented:
@bytesnz Did you find some time to integrate the things you wanted by any chance? No pressure, just making sure the PR doesn't go stale...

bytesnz commented:
@raucao I did start making changes and looking at it. I unfortunately ran out of time before going on a trip. I was going to merge the two dockerfiles and add testing to the action from peponi. I can do this at the start of Sep (long trip without a laptop).

JakubNer commented on docker/README.md:
```diff
@@ -0,0 +1,66 @@
+# armadietto [![Build Status](https://secure.travis-ci.org/remotestorage/armadietto.svg)](http://travis-ci.org/remotestorage/armadietto) [![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat-square)](https://github.com/Flet/semistandard) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/0eaafdf96ebb47a9ac462bcf6a7ccb06)](https://www. ...
+
+> ### :warning: WARNING
+> Please do not consider `armadietto` production ready, this project is still
```
Not an issue with this PR, but wondering out loud, what is the gate at which we consider Armadietto production ready?  I think hard to assess how many production workloads are already running on top of it?

raucao replied:
Good question! I think it merits its own issue or forums topic.

JakubNer commented:
@bytesnz, was testing this a bit.  Pulled your branch and...

```
docker build -t remotestorage/armadietto:latest -f .\docker\Dockerfile .
docker run -d -p 8000:8000 remotestorage/armadietto:latest
```

But it exits right away with `': No such file or directory` in the logs.

So I went into the image:

```
docker run -it --rm -p "8000:8000" remotestorage/armadietto:latest /bin/sh
```

And running the linked `/opt/armadietto/bin/armadietto.js` conks out in the same manner.

Are we certain we can start the process like that?

I've only started it as a library as per `/example/server.js`

Would love to get this out and build on top of it.  I see there are about 15k downloads of armadietto on docker-hub from various providers, would be nice to get this official one out 👍.

bytesnz commented:
Hey @JakubNer. Hmm, that's weird. Just tried running through the build process again and no issues for me (commands below). Is the image [bytesnz/armadietto:latest](https://hub.docker.com/r/bytesnz/armadietto) working for you (is built off this repo?

```bash 
git clone https://github.com/bytesnz/armadietto
cd armadietto/
git checkout  add-dockerfile 
docker build -t remotestorage/armadietto:latest -f ./docker/Dockerfile .
docker run -p 8000:8000 --name armadietto remotestorage/armadietto:latest
```

JakubNer commented:
@bytesnz, indeed when I pull your image (`bytesnz/armadietto:latest`) off docker hub, works like a charm 👍.

My own build fails like I mentioned.  I did build off of the right place:

```
PS C:\jj\src\armadietto> git status
On branch add-dockerfile
Your branch is up to date with 'bytesnz/add-dockerfile'.
```

Just rebuilt it locally, same error.

Weird.

JakubNer commented:
@bytesnz, I'm puzzled.  I checked your docker-hub (https://hub.docker.com/r/bytesnz/armadietto/tags) and the image that runs for me is the image the git actions built and pushed last Octo (https://github.com/bytesnz/armadietto/actions/runs/1385063021) which is the same commit as of this PR and your tip-of-tree in your branch.  

So I don't understand why the linked `/usr/local/bin/armadietto` invocation run inside that image doesn't execute when I re-build locally.

What always works is simply removing the link on line 44 of the `Dockerfile` and replacing line 52 with:

```
CMD node /opt/armadietto/bin/armadietto.js -c /etc/armadietto.conf.json
```

I wonder if the difference is my host.  I'm building on Windows in PowerShell.  I don't see how that should make a difference though.

raucao commented:
I thought it might help if someone else tested this to confirm, so I tried the instructions from https://github.com/remotestorage/armadietto/pull/37#issuecomment-1004570628 on Arch Linux, and it worked without issues.

However, if it works on Windows without the symlink (which I also have no idea how to explain), I guess forgoing it in favor of using the full path in the final command wouldn't hurt anyone.

JakubNer commented on docker/Dockerfile:
```diff
@@ -0,0 +1,54 @@
+RUN mkdir /opt/armadietto
+WORKDIR /opt/armadietto
+
+RUN apk add nodejs
```
Is it OK to always bring in latest version?  Should this be explicitly controlled?  I guess same for `alpine:latest` on line 1.

I'm cool with it as it is, just thinking out loud, in light of recent debacles.

Was also for indeterministic bits between October and now that could cause my failures building.  But I don't think this is it either.

raucao replied:
Not sure what the default tag there points to, but latest active LTS makes most sense to me for node.js.

DougReeder replied:
It appears that `nodejs` does refer to an LTS version

bytesnz commented:
> So I don't understand why the linked `/usr/local/bin/armadietto` invocation run inside that image doesn't execute when I re-build locally.

@JakubNer oooo that is weird. If you shell into your created docker image does the link exist and does it point to where it should?

```
/opt/armadietto $ ls -l /usr/local/bin/
total 0
lrwxrwxrwx    1 root     root            33 Oct 26 09:51 armadietto -> /opt/armadietto/bin/armadietto.js
```

I'm also curious to know what happens when you try and create a symbolic link in it

```
/opt/armadietto $ ln -s /opt/armadietto/bin/armadietto.js /tmp/
/opt/armadietto $ ls -l /tmp/
total 0
lrwxrwxrwx    1 armadiet armadiet        33 Jan 16 20:52 armadietto.js -> /opt/armadietto/bin/armadietto.js
```

But yes agreed, if the full path works, we should just switch it to that.

JakubNer commented:
>  If you shell into your created docker image does the link exist and does it point to where it should?

Indeed it exists but produces the same `': No such file or directory` output:

```
PS C:\jj\src\overhide-ledger> docker run -it --rm -p "8000:8000" remotestorage/armadietto:latest /bin/sh
/opt/armadietto $ ls -l /usr/local/bin/
total 0
lrwxrwxrwx    1 root     root            33 Jan  2 07:59 armadietto -> /opt/armadietto/bin/armadietto.js
/opt/armadietto $ armadietto
': No such file or directory
```

Relinking works and lists in `/tmp`.

raucao commented:
@bytesnz What's the status of the open tasks here? I see some unchecked ones about GitHub Actions, which I assume is about automatically building new images upon merging or for releases, right?

bytesnz commented:
In terms of the tasks:
* complete github action file (or complete later)

  Looking at now (https://github.com/bytesnz/armadietto/compare/add-dockerfile...bytesnz:expand-docker). Looking to:
  * run the tests in the action
  * allow the workflow to run on all pushes (except for the publish)

  The only issue is nicely running the e2e tests. Looking at the specs, they create in instance of armadietto themselves rather than querying an external server instance. Not sure if there is a way to run the tests on an external server, but was thinking could either:
  * mount and run the tests in the created docker container, or
  * modify the tests so that there is an option to run the tests as they are or against an external server
* add tests to github action file

  See above
* agree on Dockerfile local copy vs yarn global / npm i -g

  Silently looking for agreement/protest - currently using local copy for the benefit of being able to run the actions on non-published versions.

raucao commented:
Cool, thanks.

Not sure I understand the problem with the tests. There's already a working Action for that. Shouldn't they each to their own thing and simply kick off other Actions depending on success? I don't really know much about GitHub Actions, but pipelines composed of different workflows are pretty standard in CI these days.

bytesnz commented:
With the tests, I was thinking it would be good idea to test the docker image to ensure the image is working as expected (the docker configuration is correct and the ports/volumes are working as expected). I was thinking about using the e2e tests just as they are already written, but the test could just be a request to the running docker image to ensure it returns the expected response.

For the release, the current action will run when a release is created. The one I was working on was running on any push, but it only publishes the image when a tag is created. Can align it to the current tests (run on master pushes and release creation).

raucao commented:
> image to ensure the image is working as expected (the docker configuration is correct and the ports/volumes are working as expected).

Ah, I see now. I don't think we have to run the normal test suite for that.

> For the release, the current action will run when a release is created. The one I was working on was running on any push, but it only publishes the image when a tag is created. Can align it to the current tests (run on master pushes and release creation).

That sounds reasonable to me. If we build images for all tags, then we can test beta tags directly from the registry before tagging stable releases.

bytesnz commented:
> That sounds reasonable to me. If we build images for all tags, then we can test beta tags directly from the registry before tagging stable releases.

Yeah. Ah... just noticed there are no [tags](https://github.com/remotestorage/armadietto/tags) or [releases](https://github.com/remotestorage/armadietto/releases) at the moment. Is this something that is being looked at?

raucao commented:
Yes, definitely. We discussed Armadietto in the [monthly hangout call today](https://community.remotestorage.io/t/remotestorage-monthly-hangout-january-2022/758) and had rough consensus among the people on the call about tagging a new release soon.

@lesion I think you are the one in control of publishing to npm, right? Wondering if you want to help with the release or how we should go about it.

JakubNer commented:
@bytesnz, 

> agree on Dockerfile local copy vs yarn global / npm i -g

What does this mean?  Is this some alternative to what you did whereby we'd use's npm's registry or some such?  I don't see why we'd not want to do what you did already.

As for running tests against a running container.  Guess we'd want to add a line in `.github/workflows/docker-hub.yml ` after we run the image to run a separate chai-http test or some such?  Indeed all current mocha tests embed the server.  

Maybe that test could be a separate PR?  So it's not all just on your hands?  Piecemeal?

JakubNer commented:
@bytesnz, [made a PR](https://github.com/bytesnz/armadietto/pull/3) for your consideration.  Maybe a way forward with running the tests against the docker image.  All the tests pretty much already use chai-http so the change was small.

Unfortunately the oauth tests have some storage issue, couldn't figure it out yet.  Webfinger and file tree tests pass OK.

The PR also has a change to replace the symbolic link with a direct node `CMD` so I can run builds built on my machine (Windows).

To run, I do (powershell):

```
// first console
docker build -t remotestorage/armadietto:latest -f ./docker/Dockerfile .
docker run --rm -p 8000:8000 --name armadietto remotestorage/armadietto:latest

// second console
${env:TEST_PORT}=8000
npm run test
```

All from repo root.  So I think running tests like this should be doable from the github action workflow by just setting the `TEST_PORT` env variable.  But, need to figure out the test failures.  I didn't get any issues when creating account from the `/example/index.html` against this image.

bytesnz commented:
> What does this mean? Is this some alternative to what you did whereby we'd use's npm's registry or some such? I don't see why we'd not want to do what you did already.

@JakubNer, yeah, @peponi [suggested](https://github.com/remotestorage/armadietto/pull/37#issuecomment-846534169) using the published package for the dockerfile (quote below). I think the doing it from git will give more options in the future (eg creating docker images from non-published versions

> but this is not needed for this project IMHO because there is already a finished package on npmjs

> @bytesnz, made a PR on this branch for your consideration. Maybe a way forward with running the tests against the docker image. All the tests pretty much already use chai-http so the change was small.

Cool, yeah that looks good. Sorry I hadn't update the command already (got distracted with the [reworking the action to add the testing](https://github.com/bytesnz/armadietto/blob/expand-docker/.github/workflows/docker-hub.yml)). Have added the `CMD` change to the temp branch I'm working, which I will hopefully finish in the next few mins.

For the testing, yeah that looks good - was thinking of something similar, though was thinking we could extract the start/stop logic into its own helper file, so it only needs to be changed once. Weird about the tests not passing. Can try and have a look at some point

bytesnz commented:
Ok. So, I have ]rewritten the github action](https://github.com/bytesnz/armadietto/compare/add-dockerfile...bytesnz:expand-docker?expand=1) so that it is split into multiple jobs. This cleans it up a bit and allows tests to be run in parallel (multiple jobs).

The jobs are:
  * `version` - detects the version that should be used for tagging the docker image (GITHUB_REF which will be a tag when it is running on a release
  * `build` - build the docker image and create an artifact from the build
  * `sec_test` - runs the docker image against aquasec/trivy
  * `e2e_test` - will run the tests to make sure the image is working as expected (currently commented out)
  * `publish` - publishes the docker image (only runs on a tag [when a release is published])

Things of note:
* The `sec_test` job won't fail if it detects bad stuff... not sure if we want to enable it. Would be good to if it fails, mark it as a warning... but I am newish to actions, so not sure how to do this yet
* The action needs to be set up as a check to be helpful for PRs... again not sure how to do this yet (what can I say, I'm a gitlab user).

As it is a big change, haven't added it to this branch yet, it is currently in its own branch, but I can merge it if people are happy with the new format https://github.com/bytesnz/armadietto/compare/add-dockerfile...bytesnz:expand-docker?expand=1

JakubNer commented:
I'd vote for you to merge it.  The [pipeline runs well](https://github.com/bytesnz/armadietto/actions/runs/1769834582) and produces [an image](https://hub.docker.com/r/bytesnz/armadietto) that I just ran through the `/example/index.html` adhoc tests on local and does OK.

lesion commented:
> Yes, definitely. We discussed Armadietto in the [monthly hangout call today](https://community.remotestorage.io/t/remotestorage-monthly-hangout-january-2022/758) and had rough consensus among the people on the call about tagging a new release soon.
> 
> @lesion I think you are the one in control of publishing to npm, right? Wondering if you want to help with the release or how we should go about it.

As you may have noticed, I don't have time to track your wonderful progress in remotestorage world, so I'll be happy to transfer the package to someone else -> https://docs.npmjs.com/transferring-a-package-from-a-user-account-to-another-user-account

bytesnz commented:
> I'd vote for you to merge it. The [pipeline runs well](https://github.com/bytesnz/armadietto/actions/runs/1769834582) and produces [an image](https://hub.docker.com/r/bytesnz/armadietto) that I just ran through the `/example/index.html` adhoc tests on local and does OK.
> 
> I wouldn't worry too much that the e2e job is commented out. We can iterate.

Cool. Merged.

> As you may have noticed, I don't have time to track your wonderful progress in remotestorage world, so I'll be happy to transfer the package to someone else

Could set up a git action to do the publish to npm as well...

raucao commented:
> As you may have noticed, I don't have time to track your wonderful progress in remotestorage world, so I'll be happy to transfer the package to someone else -> https://docs.npmjs.com/transferring-a-package-from-a-user-account-to-another-user-account

Hi @lesion! Thanks for the update. Who knows, maybe you'll have time again to contribute in the future at some point.

Until then, any volunteers for maintaining the npm package? I think ideally at least two people should have permission to publish (you can invite them to the package on npm).

> Could set up a git action to do the publish to npm as well...

Yes, we could do that in addition to having more than one person with access to the npm package. This way, nobody needs to take time specifically to publish the package.

DougReeder commented:
Publishing to NPM automatically from the master branch is a good idea, if we set up a `develop` branch where PRs are merged.

If publishing to NPM needs to be done manually, I'm willing to take responsibility.

raucao commented:
> Publishing to NPM automatically from the master branch is a good idea, if we set up a `develop` branch where PRs are merged.

I think it's enough to only publish tagged releases and keep using master for normal development.

> I'm willing to take responsibility

Great! @lesion, could you add @DougReeder to the npm package then?

I think for publishing from an Action, we just need one person to obtain a token from npm, and configure it in some kind of secret variable setting on GitHub.

raucao commented:
@bytesnz Anything still missing for a merge of this PR now?

lesion commented:
> Great! @lesion, could you add @DougReeder to the npm package then?

@DougReeder let me know your npm user and I'll add you as a maintainer

DougReeder commented:
@lesion I am "dougreeder" there, also. :-)

bytesnz commented:
> @bytesnz Anything still missing for a merge of this PR now?

Only being able to tell that trivy found an issue without failing the workflow (for example there is a vulnerability with minimist that it has detected). Do we want it to fail or silently fail?

raucao commented:
I saw that same package being warned about by GitHub and closed the alert, because it's from the example folder's `package.json`, not Armadietto's actual dependencies.

Maybe we can exclude that folder from scans?

raucao commented:
> Do we want it to fail or silently fail?

I'd say that's OK for now, because we still get GitHub security warnings, as well as directly from npm installs on the command line. Thus, for npm, we don't need an extra security check in the Docker builds, but Docker security checks would be very useful, of course.

DougReeder commented on docker/Dockerfile:
```diff
@@ -0,0 +1,54 @@
+FROM alpine:latest as build
```
I'd have started from node:16-alpine, but I'm no expert on Docker.

bytesnz replied:
Yeah, using alpine as the base oddly makes for a much smaller image (double the size-ish [see https://github.com/remotestorage/armadietto/pull/37#issuecomment-846534169])

DougReeder commented on docker/Dockerfile:
```diff
@@ -0,0 +1,54 @@
+RUN apk add nodejs $PKG_MANAGER
+
+COPY package.json ./
+COPY package-lock.json ./
```
This is probably unnecessary (since yarn is used), but it doesn't hurt.

bytesnz replied:
Yeah, from the suggested changes to the file. Thought it may be useful to leave in. Shouldn't affect the build.

DougReeder commented on docker/README.md:
````diff
@@ -0,0 +1,66 @@
+## Usage
+
+```
+docker run -d -p 8000:8000 remotestorage/armadietto:latest
````
Under MacOS 11.6.2, with Docker Desktop 4.4.2, this line alone evokes the error
`Unable to find image 'remotestorage/armadietto:latest' locally`
Should there be a build step?

bytesnz replied:
Yes. Will work once it has been published, but yes, will add info about building.

bytesnz replied:
Added in [a53db93](https://github.com/remotestorage/armadietto/pull/37/commits/a53db93cb5151f7d440670d493bf7b8ed6e519a1)

bytesnz replied:
Have just added info about running the docker image behind a proxy as well as it is quite a common thing to do [723f843](https://github.com/remotestorage/armadietto/pull/37/commits/723f843bcfdaba66f4210745ef93fd60b73d3b82)

bytesnz commented:
> I'd say that's OK for now, because we still get GitHub security warnings, as well as directly from npm installs on the command line. Thus, for npm, we don't need an extra security check in the Docker builds, but Docker security checks would be very useful, of course.

Ok. Will adjust the job so the aquasec fails if it detects an issue, but the job is allowed to fail. Will result in the same, but is a little clearer.

raucao commented:
@bytesnz The Codacy badge has been removed a while ago. You might want to merge master into this branch...

bytesnz commented:
> Ok. Will adjust the job so the aquasec fails if it detects an issue, but the job is allowed to fail. Will result in the same, but is a little clearer.

Added in [55ccc56](https://github.com/remotestorage/armadietto/pull/37/commits/55ccc56a8ae55fe03d2aa7a3b3fbfa10d27635af)

![image](https://user-images.githubusercontent.com/1755886/152705099-68c56ede-7a1d-4481-9c84-8ba5b0a88b4b.png)

bytesnz commented:
> @bytesnz The Codacy badge has been removed a while ago. You might want to merge master into this branch...

Yeah, is a clean merge atm, so haven't merged it in.

DougReeder commented on docker/README.md:
```diff
@@ -24,6 +24,24 @@ docker run -d -p 8000:8000 remotestorage/armadietto:latest
 
 The armadietto will run as the armadietto user (UID 6582) in the container.
 
+### Behind a Proxy
```
It might be best to move this to the notes directory, beside the Apache notes. Reverse proxy config should be the same whether Armadietto is run using Docker or node.

bytesnz replied:
Ah yeah, could do. I was thinking this file could be the readme that is on the docker hub page for the image (eg https://hub.docker.com/repository/docker/bytesnz/armadietto), so was putting all information to get up and running. Could add a see more link to the notes folder?

raucao replied:
> Could add a see more link to the notes folder?

Good idea.

bytesnz replied:
I've updated the README ([cefdcb4](https://github.com/remotestorage/armadietto/pull/37/commits/cefdcb4787508f9d897fc1c6a99e29d56c0b995e)), cleaning the organisation up a little bit and added some references to the docs in the git repo.

DougReeder commented:
> @lesion I am "dougreeder" there, also. :-)

I have generated an NPM token, but I can't set it as a secret for the remotestorage/armadietto repo. Either I need to securely pass it to someone who can set secrets, or receive authority to add it myself.

DougReeder commented:
The "HTTPS Protection with NGINX" section on this page has some plausible-sounding advice: https://chris-knight.net/blog/2017/02/alternatives-to-fail2ban

If it's workable, it would be a fine addition to an nginx configuration page.

raucao commented:
That's a very nice blog post! But I'm not sure we should keep extending the *Docker* pull request with more information or tasks.

I think the only thing missing here are final reviews.

JakubNer commented:
> @DougReeder @JakubNer Want to do a final review and add your approval (or not)?

✔

JakubNer reviewed (approved):
Let's get this in.

DougReeder commented:
After rebasing on master, it would be nice to verify that logs can be collected outside the container, but that doesn't have to happen before this is merged.

raucao commented:
@DougReeder Want to give this a final review and approve it in case everything works and looks good?

bytesnz commented:
Sorry for the last minute. 

Was just looking at testing the merge from master and saw it may have no conflicts, but there are some breaking changes (my bad for not looking properly at it earlier). Nice update to the server pages btw.

Have done a commit to fix the changes - adding a log directory, using npm instead of yarn, update the docker config to the latest example config.

Also saw the logging has been added and added to the example config. The Dockerfile(build) is currently using a static config file *docker/config.json*. If the http port isn't going to change, I could update the Dockerfile to run `bin/armadietto.js -e` to get the latest example config. Only thing would be if the default port changed or http was disabled by default, it would break the docker image if it wasn't updated.

bytesnz commented:
@DougReeder, yeah, the `DOCKER_USER` secret will need to be set (https://github.com/remotestorage/armadietto/settings/secrets/actions) - it uses it to generate the image tag `tags: ${{ env.DOCKER_USER }}/armadietto:${{ needs.version.outputs.version }}`. The `DOCKER_TOKEN` will need to be set to publish the docker image (will only happen on a tag).

JakubNer reviewed (approved):
@bytesnz, I verified your latest merge including the new logging and new login page.

Synced your branch and...

```
docker build -t remotestorage/armadietto:latest -f .\docker\Dockerfile .
docker run -d -p 8000:8000 remotestorage/armadietto:latest
docker log ... --follow
```

Logs do show up and the new page does pop up.

BTW, I still had to change the `CMD` line in `Dockerfile` from `CMD /opt/armadietto/bin/armadietto.js -c /etc/armadietto.conf.json` to `CMD node /opt/armadietto/bin/armadietto.js -c /etc/armadietto.conf.json` 🤷‍♂️.  Regardless, I know the images work fine when github updates (from previous testing).

JakubNer commented on docker/Dockerfile:
```diff
@@ -0,0 +1,60 @@
+
+RUN ln -s /opt/armadietto/bin/armadietto.js /usr/local/bin/armadietto
+
+COPY docker/config.json /etc/armadietto.conf.json
```
Setting this up in k8s and I find myself changing `/etc` to `/usr/local/etc` to make it easier to mount my own config without clobbering the rest of the `/etc` folder:  data file mount.

Maybe we can keep this as `/usr/local/etc/..`?

bytesnz replied:
Yeah, I agree with the not clobbering */etc*. I did it based on the README, which has changed to */etc/armadietto/conf* now (so should update it in this...?). With the docker image, you can store the conf file wherever you want, it just needs to be mounted in */etc*, so it isn't too bad the way it is atm?

bytesnz commented:
> @bytesnz, I verified your latest merge including the new logging and new login page.
> 
> Synced your branch and...
> 
> ```
> docker build -t remotestorage/armadietto:latest -f .\docker\Dockerfile .
> docker run -d -p 8000:8000 remotestorage/armadietto:latest
> docker log ... --follow
> ```
> 
> Logs do show up and the new page does pop up.
> 
> BTW, I still had to change the `CMD` line in `Dockerfile` from `CMD /opt/armadietto/bin/armadietto.js -c /etc/armadietto.conf.json` to `CMD node /opt/armadietto/bin/armadietto.js -c /etc/armadietto.conf.json` 🤷‍♂️. Regardless, I know the images work fine when github updates (from previous testing).

Sorry for the delayed response. I finally got around to trying out building the docker image in windows and found the issue 😞 it's the line endings. When you check it out on windows (depending on your settings), it will change the line ending to CRLF, which, turns out, the bang line does not like. Changing the line endings back to just LF fixes the problem. 😞 
Possible solutions are:
- forcing got to not change the line ending for *bin/armadietto.js*
- ensuring the file has the right line endings in the dockerfile (by adding a `tr` or `sed` command to it)

bytesnz commented:
I've updated the Dockerfile to ensure the bin file (*bin/armadietto.js*) always has unix line endings on the docker image, so you shouldn't have anymore issues on windows @JakubNer. I've also updated the location of the conf file so it is the same as the current README.

raucao commented:
AFAICS, this PR just needs a final review from @JakubNer for Windows compatibility, correct?

DougReeder commented:
So far as I can tell, this PR could still be merged as-is.  It would be nice to merge in https://github.com/bytesnz/armadietto/pull/7/files first, though.

raucao commented:
@bytesnz Could you have a look at @DougReeder's PR to your repo?

bytesnz commented:
> @bytesnz Could you have a look at @DougReeder's PR to your repo?

done and merged :+1:

DougReeder commented:
Now we just need to merge master into this, to ensure we won't lose any of the changes to the submitBtn and documentation.

DougReeder commented:
Everyone's concerns appear to have been addressed, so I will merge this in a week, if no one objects.
